File: web/client/src/app/types/state.ts
/* 
* NOTICE
*  
* This software (or technical data) was produced for the U. S. Government
* and is subject to the Rights in Data-General Clause 52.227-14, Alt. IV
* (May 2014) – Alternative IV (Dec 2007)
* 
* (c) 2024 The MITRE Corporation. All Rights Reserved.
*/ 

import { Annotation } from './annotation';
import { JsonProperty } from '../utils/decorators/json-property';
import { ComponentRef } from '@angular/core';

export enum StateType {
  ADD,
  REMOVE,
  BULK_REMOVE,
  RESET,
  CLEAR,
  GROUP_ONLY,
  GROUP_ADD
}

export class State {
  @JsonProperty({clazz: Annotation})
  annotation: Annotation;
  data: any; // This feels like a hack
  type: StateType;
  compRef: ComponentRef<any>;

  constructor();
  constructor(annotation: Annotation, type: StateType);
  constructor(annotation?: Annotation, type?: StateType) {
    this.annotation = annotation;
    this.data = undefined;
    this.type = type;
  }
}

File: web/src/main/java/org/mitre/iwp/web/model/State.java
/* 
* NOTICE
*  
* This software (or technical data) was produced for the U. S. Government
* and is subject to the Rights in Data-General Clause 52.227-14, Alt. IV
* (May 2014) – Alternative IV (Dec 2007)
* 
* (c) 2024 The MITRE Corporation. All Rights Reserved.
*/ 

package org.mitre.iwp.web.model;

import jakarta.persistence.*;
import java.util.Map;

@Entity
public class State {
    @Id
    @GeneratedValue
    private Long id;
    public Long getId() { return this.id; }
    public void setId(Long id) { this.id = id; }

    @OneToOne(orphanRemoval = true)
    private Annotation annotation;
    public Annotation getAnnotation() { return annotation; }
    public void setAnnotation(Annotation annotation) {
        this.annotation = annotation;
    }

    @ElementCollection
    @MapKeyColumn(name="name")
    @Column(name="value")
    @CollectionTable(name="state_data_map", joinColumns={@JoinColumn(name="state_data_id", referencedColumnName = "id")})
    private Map<String, String> data;
    public Map<String, String> getData() { return data; }
    public void setData(Map<String, String> data) {
        this.data = data;
    }

    private int type;
    public int getType() { return type; }
    public void setType(int type) {
        this.type = type;
    }
}
